### ISSUE TITLE
GemStone expects Blocks for ifNotNil: and friends.  What does this code do?

### ISSUE THREAD

**Issue by pdebruic** [DuneSt/MaterialDesignLite#180]
https://github.com/DuneSt/MaterialDesignLite/blob/42ab819b19895528d3121d42e32f1b1b533a486a/src/Material-Design-Lite-Extensions.package/MDLPanelSwitcherButton.class/instance/jsOnComplete.st#L3

**Comment by jecisc** [comment]
#js is an extension we added on String and in the next version it will probably be on JSObject too to simplify code.

It does this:

```Smalltalk
js
	^ JSStream on: self
```

I can use a full block if it helps but I think it will not be the only spot where we send a symbol instead of a block to #ifNotNil:/#ifNil:/#ifTrue:/#ifFalse:/etc

I don't use gemstone. If you use MDL with Gemstone I'm open to the conversation to improve the compatibility.

**Comment by pdebruic** [comment]
So when ifNotNil: is passed a symbol it converts to 

`self perform: tehSymbol` ?

**Comment by jecisc** [comment]
In Pharo:

UndefinedObject>>ifNotNil: will do nothing. It will just return self.

ProtoObject>>ifNotNil: will do a #cull: on the parameter with self. For a block, it executes the block. For a symbol it does:

```Smalltalk
cull: anObject 
	^anObject perform: self.
```

Isn't there a Pharo compatibility package in Gemstone?

**Comment by pdebruic** [comment]
Yea but I'm not yet to the point where I want to override those messages.  I'd rather branch and change this project tbh

**Comment by jecisc** [comment]
For something as small as this we can change it in the main branches. I'll change this for the next release but I'm sure it will not be the only problem. Do not hesitate to report the problems you encounter. It would be great to have MDL working on Gemstone.

**Comment by pdebruic** [comment]
Seems like using the symbol idiom vs the block saves typing 5 characters.  And I can change it with a re-write rule in Pharo.  So unless you're opposed to a MaterialDesignLiteGemStone and MaterialDesignLitePharo package and a change to the Baseline I can do it this weekend.

**Comment by jecisc** [comment]
I think that if we want a full compatibility maybe we will need a Material-Design-Lite-Gemstone package, but I'm not sure it is necessary for this issue.  Just changing the code to 

```Smalltalk
    self onCompleteJs ifNotNil: [ :script | script js ]
```

is enough no?

I'm open to any change improving MDL feature, quality or compatibility wise if it does not add too much complexity :)

**Comment by pdebruic** [comment]
Yeah sure that will fix it for this instance for sure of the issue.  Have you used that idiom anywhere else?


I don't think it'll be too hard to either do it by hand or make the code critics or rewrite rule to just make it something we don't have to think about too much.

**Comment by jecisc** [comment]
We could also try to enable the tests for Gemstone. (I am currently writing the tests for Pharo little by little) But I doubt everything will pass since I used traits. 

I don't know if I used it somewhere else but it's highly possible. I find it more readable so I always use it if I have the choice.

**Comment by pdebruic** [comment]
Also - I noticed that the error with `#ifNotNil:` getting a Symbol vs a Block raises a compiler error when the code is loading.

**Comment by jecisc** [comment]
It might not be the only compiler error since MDL has Traits. 

I was thinking we could have Travis commit another version of each commit on master in another branch with flatten traits.

**Comment by pdebruic** [comment]
I think the GS version of monticello automatically flattens the traits.  I'm not sure though. Once the code loads I'll run through the demo webpages and see if anything dies.  

The version of MDL from 6 months ago works fine in GS

**Comment by jecisc** [comment]
I did not knew. 

The Traits were introduce in the v1.2.0. 

If it works on Gemstone we should add it to the CI and in the README.
